src/rts/PlayerAction.java

```diff
@@ -376,17 +392,19 @@ public static PlayerAction fromVectorAction(int[][] actions, GameState gs, UnitT
 				base_ru.merge(ru);
 			}
         }
+		
+		// FIXME is the clone() really necessary here? base_ru was just newly constructed,
+		// and will not be used outside of this, so I think we can just pass it directly?
         pa.setResourceUsage(base_ru.clone());
 
         for(int[] action:actions) {
             Unit u = gs.pgs.getUnitAt(action[0] % gs.pgs.width, action[0] / gs.pgs.width);
-            UnitActionAssignment uaa = gs.unitActions.get(u);
 
             // execute the action if the following happens
             // 1. The selected unit is *not* null.
             // 2. The unit selected is owned by the current player
             // 3. The unit is not currently busy (its unit action is null)
-            if (u != null && u.getPlayer() == currentPlayer && uaa == null) {
+            if (u != null && u.getPlayer() == currentPlayer && gs.unitActions.get(u) == null) {
```

Comment:
only do the map access if the previous two checks didn't fail, and avoid unnecessary local variable.